Add CLI plugins to recap

tonyskapunk opened the pull request:
- Add a new option `-p` for plugin actions in `recap`
- Lists, Enables or Disables plugins
- Moves out of recap the plugin functions
- Updates to `man`, `CHANGELOG.md` to reflect new option
- Add testing in travis

tonyskapunk committed:
Add plugin CLI support to recap

tonyskapunk committed:
Add functions to support plugin related actions

tonyskapunk committed:
Update man page to include plugin actions

tonyskapunk committed:
Update changelog with cli plugin feature

tonyskapunk committed:
Add plugin cli tests to CI

tonyskapunk commented:
@StafDehat if you have some cycles available, would you mind reviewing this? Thanks!

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+}
+
+# Removes dups from a csv string
+csv_remove_dups() {
```
This actually does more than just de-dupe a CSV.  It also turns the CSV into a space-separated list.

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+# Obtain the list of plugins in a directory
+plugins_list() {
+  local plugin_dir="$1"
+  local p_names=( $( ls -1 "${plugin_dir}" | tr '\n' ' ') )
```
Heads-up, using space as a delimiter here means that any plugin with a space in its filename, will break this function.  That might be acceptable, but if you want to allow it, you'll likely have to move away from arrays a bit, and rely on newline-separated strings.

tonyskapunk replied:
Your comment is correct.  I know this is an *assumption* but any plugin with spaces in its name is not expected to work, it has been documented in here: https://github.com/rackerlabs/recap/blob/master/README.md#naming-conventions

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+    "disabled")
+      plugins_a=( $( plugins_list "${PLUGIN_A_DIR}" ) )
+      plugins_e=( $( plugins_list "${PLUGIN_E_DIR}" ) )
+      # Loop through the available plugins
```
I'd encourage you to use 'comm' or 'diff' here instead of copying an array within a loop.  I think it'd be more efficient, and to me it seems simpler to read.  In addition, if you move away from whitespace-separated lists in the plugins_list() output and use newline-separated lists instead, it'll already be in the right format:
```
# Show plugins in the Available dir, but not in Enabled:
comm -23 <( plugins_list "${PLUGIN_A_DIR}" ) <( plugins_list "${PLUGIN_E_DIR}" )
```

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+}
+
+# Obtain a list of plugin by status
+get_plugin_status() {
```
We should rename this function - get_plugin_status() sounds to me like it'll take a plugin name as an argument, and tell you what its status is.  Call this get_plugins_by_status() or something, instead.

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+}
+
+# Prints the status of plugins
+print_plugins() {
```
This function name should be tweaked too - it suggests it'll print everything, when in actually it takes a status as argument, and prints only the plugins in that status.  Which... sounds a heck of a lot like the function right above this one.  These two functions could likely be simplified into one.

tonyskapunk replied:
Ah, right, I'm getting rid of the `print_plugins` and re-using the above (newly renamed) function `get_plugin_by_status`, also moving the validation into a new function `validate_plugin_status`

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+   local recap_conf="/etc/recap.conf"
+
+   # Only perform changes when not already defined
+   if [[ "${USEPLUGINS}" == "no" ]]; then
```
To reduce nesting, test for USEPLUGINS=yes, and return.  Then the rest of this function can be 1-less-nest.

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+}
+
+# Enable Plugin support in recap.conf
+enable_plugin_conf() {
```
Seems like the goal of this whole function is to ensure that recap.conf contains exactly 1 line setting USEPLUGINS, and that it sets it to "yes".  As such, this function seems excessively complicated.
I propose this instead:
```
# If already enabled, just return:
[[ "${USEPLUGINS}" == "yes" ]] && return

# If config already mentions USEPLUGINS, ensure they're all commented:
if grep -qP '^\s*(#.*)?\s*USEPLUGINS\s*=' "${recap_conf}"; then
  # Comment all of 'em:
  sed -i '/^\s*USEPLUGINS\s*=/s/^/#/' "${recap_conf}"
fi

# Add one entry at the bottom to enable USEPLUGINS:
echo 'USEPLUGINS="yes"' >> "${recap_conf}"
```

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+    if in_array ${plugin} ${p_a[@]}; then
+      local symlink="${PLUGIN_E_DIR}/${plugin}"
+
+      # Enable disabled plugin
```
I think we could increase readability by abstracting the actual filesystem operations within "enable_plugin" and "disable_plugin" helper functions.

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+    fi
+  done
+
+  # Print what plugins where enabled/disabled if needed
```
Typo: s/where/were/

StafDehat commented on src/core/plugins:
```diff
@@ -0,0 +1,306 @@
+        }
+      fi
+      actioned+=( ${plugin} )
+    fi
```
This would be a good spot to log debug info about skipping a plugin, due to it already being in the desired state.

tonyskapunk committed:
Rename, Add and delete functions per feedback.

tonyskapunk commented:
Thanks for all the great feedback @StafDehat !

Coming commit should cover most the feedback given.

Some additional changes were required as well, here a gist of the changes:

Renamed few functions:
- `csv_remove_dups()` --> `csv_to_list()`
- `plugin_validate_action()` --> `validate_plugin_action()`
- `get_plugin_status()` --> `get_plugin_by_status()`
- `validate_plugin()`--> `validate_plugin_name()`

Newly added:
- `validate_plugin_status()`
- `enable_plugin()`
- `disable_plugin()`

Deleted:
- `print_plugins()`

---

The only pending change is `enable_plugin_conf()` from your feedback: https://github.com/rackerlabs/recap/pull/224#discussion_r302747782 you're absolutely right, is overcomplicated, and yes, the end goal is to have a single config in it when enabling plugins.  The complexity comes when attempting to add the config in a particular place and avoiding adding additional entries.  I honestly like your suggestion a lot more, because is easy, let me think just any reasons why that not be good and then I'll submit that change as well.

tonyskapunk committed:
Simplify plugin enable function per feedback

tonyskapunk commented:
Including now your suggestion, thanks! 

I just modified a bit the query regex, since we don't care about those already commented out, i.e.

with:  `^\s*(#.*)?\s*USEPLUGINS\s*=`
```
$ echo '
> USEPLUGINS="no"
> USEPLUGINS="no" # comment
>   USEPLUGINS="no"
>   USEPLUGINS="no" #comment
> #comment   USEPLUGINS="no"
> # comment   USEPLUGINS="no"
>   # comment   USEPLUGINS="no"
> '  | grep -P '^\s*(#.*)?\s*USEPLUGINS\s*='
USEPLUGINS="no"
USEPLUGINS="no" # comment
  USEPLUGINS="no"
  USEPLUGINS="no" #comment
#comment   USEPLUGINS="no"
# comment   USEPLUGINS="no"
  # comment   USEPLUGINS="no"
``` 

with: `^\s*USEPLUGINS='
```
$ echo '
USEPLUGINS="no"
USEPLUGINS="no" # comment
  USEPLUGINS="no"
  USEPLUGINS="no" #comment
#comment   USEPLUGINS="no"
# comment   USEPLUGINS="no"
  # comment   USEPLUGINS="no"
'  | grep -P '^\s*USEPLUGINS='
USEPLUGINS="no"
USEPLUGINS="no" # comment
  USEPLUGINS="no"
  USEPLUGINS="no" #comment
```

tonyskapunk commented:
I'll go ahead and merge this one, I've implemented all the suggested changes and is doing what's expected. Can always revisit prior merging into master.
